lib/handler/cache-handler.js

```diff
@@ -0,0 +1,379 @@
+    ) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+    if (!canCacheResponse(statusCode, headers, cacheControlDirectives)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#value = {
+        complete: false,
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        body: [],
+        vary: varyDirectives,
+        size: currentSize,
+        cachedAt: now,
+        staleAt: now + staleAt,
+        deleteAt: now + deleteAt
+      }
+    }
+
+    if (typeof this.#handler.onHeaders === 'function') {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+  }
+
+  /**
+   * @see {DispatchHandlers.onData}
+   *
+   * @param {Buffer} chunk
+   * @returns {boolean}
+   */
+  onData (chunk) {
+    if (this.#value) {
+      this.#value.size += chunk.length
+
+      if (this.#value.size > this.#getMaxEntrySize()) {
+        this.#value = null
+      } else {
+        this.#value.body.push(chunk)
+      }
+    }
+
+    if (typeof this.#handler.onData === 'function') {
+      return this.#handler.onData(chunk)
+    }
+  }
+
+  /**
+   * @see {DispatchHandlers.onComplete}
+   *
+   * @param {string[] | null} rawTrailers
+   */
+  onComplete (rawTrailers) {
+    if (this.#value) {
+      this.#value.complete = true
+      this.#value.rawTrailers = rawTrailers
+      this.#value.size += this.#getSizeOfBuffers(rawTrailers)
+
+      // If we're still under the max entry size, let's add it to the cache
+      if (this.#getMaxEntrySize() > this.#value.size) {
+        const result = this.#opts.store.put(this.#req, this.#value)
+        if (result && result.constructor.name === 'Promise') {
+          result.catch(err => this.#handler.onError(err))
+        }
+      }
+    }
+
+    if (typeof this.#handler.onComplete === 'function') {
+      return this.#handler.onComplete(rawTrailers)
+    }
+  }
+
+  /**
+   * @see {DispatchHandlers.onError}
+   *
+   * @param {Error} err
+   */
+  onError (err) {
+    this.#value = undefined
+    if (typeof this.#handler.onError === 'function') {
+      this.#handler.onError(err)
+    }
```

Comment:
this would always be set

Reply:
The types allow it to be undefined however? https://github.com/nodejs/undici/blob/0ad6007ae74ce2d392c76a90b07e6cda725d3bad/types/dispatcher.d.ts#L220

Reply:
Yeah, not sure why's that but I haven't find that scenario at all

Reply:
Since it's allowed in the type I think we should check for it still